instance/models/server.py

```
@@ -0,0 +1,274 @@
+            if isinstance(target_status, str):
+                if self.status == target_status:
+                    break
+            else: # Status list
+                if self.status in target_status:
+                    break
+            time.sleep(1)
+        return self.status
+
+    @staticmethod
+    def on_post_save(sender, instance, created, **kwargs):
+        """
+        Called when an instance is saved
+        """
+        publish_data('notification', {
+            'type': 'server_update',
+            'server_pk': instance.pk,
+        })
+
+    def update_status(self, provisioned=False, rebooting=False):
+        """
+        Check the current status and update it if it has changed
+        """
+        raise NotImplementedError
+
+
+class OpenStackServer(Server):
+    """
+    A Server VM hosted on an OpenStack cloud
+    """
+    instance = models.ForeignKey(OpenEdXInstance, related_name='server_set')
+    openstack_id = models.CharField(max_length=250, db_index=True, blank=True)
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.nova = openstack.get_nova_client()
+
+    def __str__(self):
+        if self.openstack_id:
+            return self.openstack_id
+        else:
+            return 'New OpenStack Server'
+
+    @property
+    def os_server(self):
+        """
+        OpenStack nova server API endpoint
+        """
+        if not self.openstack_id:
+            assert self.status == 'new'
+            self.start()
+        return self.nova.servers.get(self.openstack_id)
+
+    @property
+    def public_ip(self):
+        """
+        Return one of the public address(es)
+        """
+        if not self.openstack_id:
+            return None
+
+        public_addr = openstack.get_server_public_address(self.os_server)
+        if not public_addr:
+            return None
+
+        return public_addr['addr']
+
+    def update_status(self, provisioned=False, rebooting=False):
```

@antoviaque nit: should probably raise exceptions for invalid state transitions (i.e. `provisioned=True` and `status != 'booted'`)